### PR TITLE
feat: Support HF16

### PR DESCRIPTION
**Pull request by thanhdanh**
# Summary
- Add `pre_step2_tie_unspent_outs_to_mix_outs_for_all_future_tx_attempts` function
- Tweaking the mechanism to set bp_version in rct_config in create_transaction function
Closes: [ticket](https://app.asana.com/0/0/1202765744038599/f)

### CODE REVIEW

File: src/monero_transfer_utils.cpp [hunk 1]
```diff
@@ -350,6 +365,72 @@ void monero_transfer_utils::send_step1__prepare_params_for_get_decoys(
...
+	Tie_Outs_to_Mix_Outs_RetVals &retVals,
+	//
+	const vector<SpendableOutput> &using_outs,
+	vector<RandomAmountOutputs> mix_outs_from_server,
+	//
+	const optional<SpendableOutputToRandomAmountOutputs> &prior_attempt_unspent_outs_to_mix_outs
+) {
+	retVals.errCode = noError;
+	//
+	// combine newly requested mix outs returned from the server, with the already known decoys from prior tx construction attempts,
```
Comment (feri42) [inline src/monero_transfer_utils.cpp:379]: What server are we talking about here?

File: src/serial_bridge_utils.hpp [hunk 2]
```diff
@@ -82,11 +82,11 @@ namespace serial_bridge_utils
 	//
 	// JSON keys - Ret vals
 	// - - Error
-	static inline string ret_json_key__any__err_msg() { return "err_msg"; } // optional
+	static inline string ret_json_key__any__err_msg() { return "err_msg"; }	  // optional
```
Comment (feri42) [inline src/serial_bridge_utils.hpp]: nit: unnecessary space added
```suggestion
	static inline string ret_json_key__any__err_msg() { return "err_msg"; } // optional
```

File: src/serial_bridge_utils.hpp [hunk 3]
```diff
@@ -82,11 +82,11 @@ namespace serial_bridge_utils
...
 	static inline string ret_json_key__any__err_code() { return "err_code"; } // optional
 	//
 	// - - Shared
-	static inline string ret_json_key__generic_retVal() { return "retVal"; } 
+	static inline string ret_json_key__generic_retVal() { return "retVal"; }
```
Comment (feri42) [inline src/serial_bridge_utils.hpp]: nit: unnecessary refactor
```suggestion
	static inline string ret_json_key__generic_retVal() { return "retVal"; } 
```

**Comment by feri42** [comment]
Can this PR be split into:
1. merge upstream changes (a lot of it is formatting)
2. our changes on top of that

**Review by gutenye — APPROVED** [review]
utACK

**Comment by thanhdanh** [comment]
> Can this PR be split into:
> 
> 1. merge upstream changes (a lot of it is formatting)
> 2. our changes on top of that

Some changes about formatting are caused by my IDE which is formatting on saving.  Let me revert if it's useless.

**Comment by feri42** [comment]
> Some changes about formatting are caused by my IDE which is formatting on saving. Let me revert if it's useless.

We want to keep the original formatting. Otherwise it becomes increasingly more difficult to compare changes on the two repos as the upstream repo advances.

**Review by feri42 — APPROVED** [review]
utACK

**Comment by feri42** [comment]
post-merge utACK

**Pull request merged**